staging/src/k8s.io/component-base/metrics/counter.go

```diff
@@ -106,28 +105,17 @@ func (c *Counter) initializeDeprecatedMetric() {
 }
 
 // WithContext allows the normal Counter metric to pass in context.
-func (c *Counter) WithContext(ctx context.Context) CounterMetric {
-	c.ctx = ctx
-	return c.CounterMetric
-}
-
-// withExemplar initializes the exemplarMetric object and sets the exemplar value.
-func (c *Counter) withExemplar(v float64) {
-	(&exemplarCounterMetric{c}).withExemplar(v)
-}
-
-func (c *Counter) Add(v float64) {
-	c.withExemplar(v)
-}
-
-func (c *Counter) Inc() {
-	c.withExemplar(1)
+func (c *Counter) WithContext(ctx context.Context) *CounterWithContext {
+	return &CounterWithContext{
+		ctx:     ctx,
+		Counter: c,
+	}
 }
 
 // withExemplar attaches an exemplar to the metric.
-func (e *exemplarCounterMetric) withExemplar(v float64) {
-	if m, ok := e.CounterMetric.(prometheus.ExemplarAdder); ok {
-		maybeSpanCtx := trace.SpanContextFromContext(e.ctx)
+func (c *CounterWithContext) withExemplar(v float64) {
+	if m, ok := c.CounterMetric.(prometheus.ExemplarAdder); ok {
```

Comment:
do we really need to check that every time?

Wouldn't it be better to implement `AddWithExemplar`?